### DISCUSSION TITLE
Supervisor Emergency Landing

### DISCUSSION THREAD

**Discussion by TommiAux** [bitcraze/discussions#923] [Q&A]
Hi,

recently I updated my local crazyflie-firmware repository and now I am having a hard time understanding the new supervisor state machine. Previously, I added emergency landing functionality (using the high level commander) in the commander module, in the case the connection was lost. This was pretty straight forward and worked perfectly fine

```diff
void commanderGetSetpoint(setpoint_t *setpoint, const state_t *state)
{
 xQueuePeek(setpointQueue, setpoint, 0);
  lastUpdate = setpoint->timestamp;
  uint32_t currentTime = xTaskGetTickCount();

+  if ((currentTime - setpoint->timestamp) > COMMANDER_WDT_TIMEOUT_SHUTDOWN) {
+    if(supervisorIsFlying()) {
+      crtpCommanderHighLevelLand(0.2, 4.0);
    } else {
      memcpy(setpoint, &nullSetpoint, sizeof(nullSetpoint));
    }
  } else if ((currentTime - setpoint->timestamp) > COMMANDER_WDT_TIMEOUT_STABILIZE) {
    ...
  }
}
```

Now the supervisor seems to be responsible for detecting connection loss. I tried to modify the `supervisorOverrideSetpoint` function in two different ways. By invoking the high level commander when the state machine gets into `supervisorStateWarningLevelOut` and by adding a new query for the state `supervisorStateExceptFreeFall` and also trying to invoke the high level commander's landing function. However, the Crazyflies still plummet to the ground after a disconnect.

```diff
void supervisorOverrideSetpoint(setpoint_t* setpoint) {
  SupervisorMem_t* this = &supervisorMem;
  switch(this->state){
    case supervisorStateReadyToFly:
      // Fall through
    case supervisorStateFlying:
      // Do nothing
      break;

    case supervisorStateWarningLevelOut:
+   crtpCommanderHighLevelLand(0.2, 4.0);
-    setpoint->mode.x = modeDisable;
-    setpoint->mode.y = modeDisable;
-    setpoint->mode.roll = modeAbs;
-    setpoint->mode.pitch = modeAbs;
-    setpoint->mode.yaw = modeVelocity;
-    setpoint->attitude.roll = 0;
-    setpoint->attitude.pitch = 0;
-    setpoint->attitudeRate.yaw = 0;
-    // Keep Z as it is
      break;

+    case supervisorStateExceptFreeFall:
+     crtpCommanderHighLevelLand(0.2, 4.0);
+    break;

    default:
      // Replace with null setpoint to stop motors
      memcpy(setpoint, &nullSetpoint, sizeof(nullSetpoint));
      break;
  }
}
```

If someone has a hint where to put such an emergency landing functionality, I'd be very grateful :)

Thanks in advance
Tommi

**Comment by krichardsson** [comment]
As soon as the state machine enters `supervisorStateExceptFreeFall` [(or actually leaves `supervisorStateWarningLevelOut`)](https://github.com/bitcraze/crazyflie-firmware/blob/89ed3225d62499c529fe72f7b0362ea0e759df6f/src/modules/src/supervisor.c#L362-L367)
[`areMotorsAllowedToRun`](https://github.com/bitcraze/crazyflie-firmware/blob/89ed3225d62499c529fe72f7b0362ea0e759df6f/src/modules/src/stabilizer.c#L323-L327) will be false and stop the motors.

The proper way to implement this would probably be to add a new state that handles the landing, maybe [between `Warning level out` and `Exception free fall`](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/supervisor/states/)?
